File: example/javalib/springboot/3-todo-spring-boot/src/com/example/todoitem/web/TodoItemController.java
package com.example.todoitem.web;

import com.example.todoitem.TodoItem;
import com.example.todoitem.TodoItemNotFoundException;
import com.example.todoitem.TodoItemRepository;
import java.util.List;
import java.util.stream.Collectors;
import javax.validation.Valid;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

@Controller
@RequestMapping("/")
public class TodoItemController {

  private final TodoItemRepository repository;

  public TodoItemController(TodoItemRepository repository) {
    this.repository = repository;
  }

  @GetMapping
  public String index(Model model) {
    addAttributesForIndex(model, ListFilter.ALL);
    return "index";
  }

  @GetMapping("/active")
  public String indexActive(Model model) {
    addAttributesForIndex(model, ListFilter.ACTIVE);
    return "index";
  }

  @GetMapping("/completed")
  public String indexCompleted(Model model) {
    addAttributesForIndex(model, ListFilter.COMPLETED);
    return "index";
  }

  private void addAttributesForIndex(Model model, ListFilter listFilter) {
    model.addAttribute("item", new TodoItemFormData());
    model.addAttribute("todoFilter", listFilter);
    model.addAttribute("todoItems", getTodoItems(listFilter));
    model.addAttribute("totalItemCount", repository.count());
    model.addAttribute("activeItemCount", getactiveItemCount());
    model.addAttribute("completedItemCount", getcompletedItemCount());
  }

  @PostMapping
  public String addNewTodoItem(@Valid @ModelAttribute("item") TodoItemFormData formData) {
    repository.save(new TodoItem(formData.getTitle(), false));

    return "redirect:/";
  }

  @PutMapping("/{id}/toggle")
  public String toggleSelection(@PathVariable("id") Long id) {
    TodoItem todoItem =
        repository.findById(id).orElseThrow(() -> new TodoItemNotFoundException(id));

    todoItem.setCompleted(!todoItem.isCompleted());
    repository.save(todoItem);
    return "redirect:/";
  }

  @PutMapping("/toggle-all")
  public String toggleAll() {
    List<TodoItem> todoItems = repository.findAll();
    for (TodoItem todoItem : todoItems) {
      todoItem.setCompleted(!todoItem.isCompleted());
      repository.save(todoItem);
    }
    return "redirect:/";
  }

  @DeleteMapping("/{id}")
  public String deleteTodoItem(@PathVariable("id") Long id) {
    repository.deleteById(id);

    return "redirect:/";
  }

  @DeleteMapping("/completed")
  public String deleteCompletedItems() {
    List<TodoItem> items = repository.findAllByCompleted(true);
    for (TodoItem item : items) {
      repository.deleteById(item.getId());
    }
    return "redirect:/";
  }

  private List<TodoItemDto> getTodoItems(ListFilter filter) {
    return switch (filter) {
      case ALL -> convertToDto(repository.findAll());
      case ACTIVE -> convertToDto(repository.findAllByCompleted(false));
      case COMPLETED -> convertToDto(repository.findAllByCompleted(true));
    };
  }

  private List<TodoItemDto> convertToDto(List<TodoItem> todoItems) {
    return todoItems.stream()
        .map(todoItem ->
            new TodoItemDto(todoItem.getId(), todoItem.getTitle(), todoItem.isCompleted()))
        .collect(Collectors.toList());
  }

  private int getactiveItemCount() {
    return repository.countAllByCompleted(false);
  }

  private int getcompletedItemCount() {
    return repository.countAllByCompleted(true);
  }

  public static record TodoItemDto(long id, String title, boolean completed) {}

  public enum ListFilter {
    ALL,
    ACTIVE,
    COMPLETED
  }
}

File: example/kotlinlib/web/7-todo-spring-boot/src/com/example/todoitem/web/TodoItemController.kt
package com.example.todoitem.web

import com.example.todoitem.TodoItem
import com.example.todoitem.TodoItemNotFoundException
import com.example.todoitem.TodoItemRepository
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.*
import javax.validation.Valid

@Controller
@RequestMapping("/")
class TodoItemController(private val repository: TodoItemRepository) {

    @GetMapping
    fun index(model: Model): String {
        addAttributesForIndex(model, ListFilter.ALL)
        return "index"
    }

    @GetMapping("/active")
    fun indexActive(model: Model): String {
        addAttributesForIndex(model, ListFilter.ACTIVE)
        return "index"
    }

    @GetMapping("/completed")
    fun indexCompleted(model: Model): String {
        addAttributesForIndex(model, ListFilter.COMPLETED)
        return "index"
    }

    private fun addAttributesForIndex(model: Model, listFilter: ListFilter) {
        model.addAttribute("item", TodoItemFormData())
        model.addAttribute("todoFilter", listFilter)
        model.addAttribute("todoItems", getTodoItems(listFilter))
        model.addAttribute("totalItemCount", repository.count())
        model.addAttribute("activeItemCount", getActiveItemCount())
        model.addAttribute("completedItemCount", getCompletedItemCount())
    }

    @PostMapping
    fun addNewTodoItem(@Valid @ModelAttribute("item") formData: TodoItemFormData): String {
        repository.save(TodoItem(id = null, title = formData.title, completed = false))
        return "redirect:/"
    }

    @PutMapping("/{id}/toggle")
    fun toggleSelection(@PathVariable("id") id: Long): String {
        val todoItem = repository.findById(id).orElseThrow { TodoItemNotFoundException(id) }
        todoItem.completed = !todoItem.completed
        repository.save(todoItem)
        return "redirect:/"
    }

    @PutMapping("/toggle-all")
    fun toggleAll(): String {
        val todoItems = repository.findAll()
        todoItems.forEach { it.completed = !it.completed }
        repository.saveAll(todoItems)
        return "redirect:/"
    }

    @DeleteMapping("/{id}")
    fun deleteTodoItem(@PathVariable("id") id: Long): String {
        repository.deleteById(id)
        return "redirect:/"
    }

    @DeleteMapping("/completed")
    fun deleteCompletedItems(): String {
        val items = repository.findAllByCompleted(true)
        items.forEach { repository.deleteById(it.id) }
        return "redirect:/"
    }

    private fun getTodoItems(filter: ListFilter): List<TodoItemDto> = when (filter) {
        ListFilter.ALL -> convertToDto(repository.findAll())
        ListFilter.ACTIVE -> convertToDto(repository.findAllByCompleted(false))
        ListFilter.COMPLETED -> convertToDto(repository.findAllByCompleted(true))
    }

    private fun convertToDto(todoItems: List<TodoItem>): List<TodoItemDto> = todoItems.map {
        TodoItemDto(
            id = it.id ?: throw IllegalArgumentException("TodoItem ID cannot be null"),
            title = it.title,
            completed = it.completed,
        )
    }

    private fun getActiveItemCount(): Int = repository.countAllByCompleted(false)

    private fun getCompletedItemCount(): Int = repository.countAllByCompleted(true)

    data class TodoItemDto(val id: Long, val title: String, val completed: Boolean)

    enum class ListFilter {
        ALL,
        ACTIVE,
        COMPLETED,
    }
}
